sqlflow_models/score_card.py

```diff
@@ -0,0 +1,104 @@
+#!/bin/env python
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.python.data import make_one_shot_iterator
+from tensorflow.keras.losses import kld
+from tensorflow.keras.optimizers import SGD
+import numpy as np
+import pandas as pd
+import scipy.stats.stats as stats
+import sklearn
+from sklearn.linear_model import LogisticRegression
+from sklearn.model_selection import train_test_split
+from sklearn.metrics import roc_auc_score, auc
+import pickle
+
+
+def optimizer():
+    return None
+
+
+def loss():
+    return None
+
+
+class ScoreCard(keras.Model):
+
+    def __init__(self, feature_columns=None, pf_bin_size=5):
+        super(ScoreCard, self).__init__(name='ScoreCard')
+
+        self._factor = 20/np.log(2)
+        self._offset = 600 - 20*np.log(20) / np.log(2)
+        self._bins = dict()
+        self._pf_bin_size = pf_bin_size
+
+    def _pf_bin(self, y, x):
+        # population frequency bucket
+        bad_num = y.sum()
+        good_num = y.count() - y.sum()
+        d1 = pd.DataFrame({'x': x,'y': y,'bucket': pd.qcut(x, self._pf_bin_size, duplicates='drop')})
+        d2 = d1.groupby('bucket',as_index=True)
+        d3 = pd.DataFrame(d2.x.min(),columns=['min_bin']) 
+
+        d3["min"] = d2.min().x
+        d3["max"] = d2.max().x
+        d3["badcostum"] = d2.sum().y
+        d3["goodcostum"] = d2.count().y - d2.sum().y
+        d3["total"] = d2.count().y
+        d3["bad_rate"] = d2.sum().y/d2.count().y
+        d3["woe"] = np.log(d3["badcostum"]/d3["goodcostum"]*good_num/bad_num)
+        iv = ((d3["badcostum"]/bad_num-d3["goodcostum"]/good_num)*d3["woe"])
+        d3["iv"] = iv
+        woe = list(d3["woe"].round(6))
+        cut = list(d3["max"].round(6))
+        cut.insert(0, float("-inf"))
+        cut[-1] = float("inf")
+        return d3, cut, woe, iv
+
+    def _to_dataframe(self, dataset):
+        x_df = pd.DataFrame()
+        y_df = pd.DataFrame()
+        for _, minibatch in enumerate(dataset):
+            data, label = minibatch
+            dx = {}
+            dy = {}
+            for name, value in data.items():
+                dx[name] = value.numpy()[0][0]
+            x_df = x_df.append(dx, ignore_index=True)
+            dy['label'] = label.numpy()[0]
+            y_df = y_df.append(dy, ignore_index=True)
+        return x_df, y_df
+
+    def _replace_woe(self, x, cut, woe):
+        return pd.cut(x, cut, labels=pd.Categorical(woe))
+
+    def _woe_encoder(self, x, y):
+        x_train_dict = {}
+        for col in x.columns:
+            dfx, cut, woe, iv = self._pf_bin(y, x[col])
+            self._bins[col] = (dfx, cut, woe, iv)
+            # replacing by the WOE encode
+            x_train_dict[col] = self._replace_woe(x[col], cut, woe)
+        return pd.DataFrame.from_dict(x_train_dict)
+
+    def sqlflow_train_loop(self, dataset, epochs=1, verbose=0):
```

Comment:
Also define `sqlflow_predict_one` to do prediction later?

Reply:
Yes will add it if needed.The training process would print the score card as follows, 

``` text
 THE SCORE CARD:
 age (20.999, 38.0] 3.0
 age (38.0, 47.0] 5.0
 age (47.0, 56.0] -1.0
 age (56.0, 65.0] -3.0
 age (65.0, 97.0] -9.0
 debt_ratio (-0.001, 0.131] -11.0
 debt_ratio (0.131, 0.286] -0.0
 debt_ratio (0.286, 0.474] -7.0
 debt_ratio (0.474, 3.248] 11.0
 debt_ratio (3.248, 15466.0] -1.0
 monthly_income (-0.001, 3325.8] 3.0
 monthly_income (3325.8, 5212.6] 5.0
 monthly_income (5212.6, 5400.0] -3.0
 monthly_income (5400.0, 8169.4] -3.0
 monthly_income (8169.4, 208333.0] -4.0
 number_of_dependents (-0.001, 2.0] 0.0
 number_of_dependents (2.0, 8.0] -3.0
 number_of_open_credit_lines_and_loans (-0.001, 4.0] 4.0
 number_of_open_credit_lines_and_loans (4.0, 6.0] -1.0
 number_of_open_credit_lines_and_loans (6.0, 9.0] -4.0
 number_of_open_credit_lines_and_loans (9.0, 12.0] -4.0
 number_of_open_credit_lines_and_loans (12.0, 46.0] 4.0
 number_of_time30_59days_past_due_not_worse (-0.001, 98.0] 0.0
 number_of_time60_89_days_past_due_not_worse (-0.001, 98.0] 0.0
 number_of_times_90_days_late (-0.001, 98.0] 0.0
 number_real_estate_loans_or_lines (-0.001, 1.0] -2.0
 number_real_estate_loans_or_lines (1.0, 2.0] -1.0
 number_real_estate_loans_or_lines (2.0, 9.0] 18.0
 revolving_utilization_of_unsecured_lines (-0.001, 0.0188] -35.0
 revolving_utilization_of_unsecured_lines (0.0188, 0.0866] -44.0
 revolving_utilization_of_unsecured_lines (0.0866, 0.273] -39.0
 revolving_utilization_of_unsecured_lines (0.273, 0.686] -10.0
 revolving_utilization_of_unsecured_lines (0.686, 2340.0] 43.0
```